tools/openslide-write-associated-images.1.in
.\"
.\" OpenSlide, a library for reading whole slide image files
.\"
.\" Copyright (c) 2007-2012 Carnegie Mellon University
.\" Copyright (c) 2023 Alexandr Virodov
.\" All rights reserved.
.\"
.\" OpenSlide is free software: you can redistribute it and/or modify
.\" it under the terms of the GNU Lesser General Public License as
.\" published by the Free Software Foundation, version 2.1.
.\"
.\" OpenSlide is distributed in the hope that it will be useful,
.\" but WITHOUT ANY WARRANTY; without even the implied warranty of
.\" MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
.\" GNU Lesser General Public License for more details.
.\"
.\" You should have received a copy of the GNU Lesser General Public
.\" License along with OpenSlide. If not, see
.\" <http://www.gnu.org/licenses/>.
.\"


.\" See man-pages(7) for formatting conventions.


.TH OPENSLIDE-WRITE-ASSOCIATED-IMAGES 1 2023-05-09 "OpenSlide @SUFFIXED_VERSION@" "User Commands"

.mso www.tmac

.SH NAME
openslide-write-associated-images \- Query or write associated images  of a virtual slide.

.SH SYNOPSIS
.BR "openslide-write-associated-images " [ --help "] [" --version ]
.I slide-file - to list associated images.
.I slide-file associated_image output.png - to write an associated image.

.SH DESCRIPTION
Query a list of associated images, or write a single associated image to a PNG file.

.SH OPTIONS
.TP
.B --help
Display usage summary.

.TP
.B --version
Display version and copyright information.

.SH EXIT STATUS
.B openslide-write-associated-images
returns 0 on success, 1 if an error occurred, or 2 if the arguments are
invalid.

.SH COPYRIGHT
Copyright \(co 2007-2023 Carnegie Mellon University and others

OpenSlide is free software: you can redistribute it and/or modify it under
the terms of the
.URL http://gnu.org/licenses/lgpl-2.1.html "GNU Lesser General Public License, version 2.1" .

OpenSlide comes with NO WARRANTY, to the extent permitted by law.  See the
GNU Lesser General Public License for more details.

.SH SEE ALSO
.BR openslide-quickhash1sum (1),
.BR openslide-show-properties (1)


CHANGELOG.md
# Notable Changes in OpenSlide


## Version 4.0.0, 2023-10-11

### Breaking changes

* Update soname to `libopenslide.so.1`
* Remove all deprecated functions
* Clear `openslide_read_associated_image()` output buffer on error
* Remove undocumented NULL `dest` support in `openslide_read_associated_image()`
* Change GLib log domain to `OpenSlide`
* Convert build system to Meson (thanks, Jan Harkes)

### New features

* New format: DICOM WSI (thanks, John Cupitt and Jim O'Donnell)
* Add APIs to read ICC color profiles (thanks, John)
* Add APIs to configure tile cache size and share caches between slides
* Add properties for associated image metadata
* generic-tiff: Set MPP properties if available
* philips: Set objective power property if available
* Add `slidetool` command-line tool which supports all OpenSlide features
* Combine all command-line tools into the same binary
* Add ICC profile to PNG images written by command-line tools
* Add self-test that doesn't require sample data (run with `meson test`)

### Changes

* Require libtiff ≥ 4, OpenJPEG ≥ 2.1, GLib ≥ 2.56, SQLite ≥ 3.14
* Require libdicom ≥ 1.0, with build-time fallback for now
* Remove support for including `openslide.h` in Visual Studio \< 2013
* Fail `openslide_open()` with a broken pixman 0.38.x
* Avoid extra buffer copy in `openslide_read_region()`
* Replace `goto`-based cleanup with `g_autoptr`
* Stop using deprecated GLib slice allocator
* Use internal wrappers for file I/O
* Documentation improvements

### Bug fixes

* Use UTF-8 filenames on Windows
* Improve `openslide-write-png` performance for very large regions
* Fix assertions on JPEG decode errors when compiled with Clang
* Portability fixes (thanks, Billy Robert O'Neal III and Kleis Auke Wolthuizen)
* aperio: Set objective power property even if floating point
* hamamatsu: Fix `Restart marker not found` on VMS slides with multiple Z-layers
* hamamatsu: Fix integer overflow in VMS parsing (thanks, Adam Goode)
* mirax: Fix `Expected 1 value` error
* philips: Don't cache missing tiles
* sakura: Fix memory leak reading missing tile


## Version 3.4.1, 2015-04-20

* New formats: Philips TIFF, Ventana TIFF
* Support OpenJPEG 2.1.0
* Improve performance of JPEG and JP2K decoding
* Add `openslide.region[i].*` properties
* Improve MATLAB compatibility
* Enable function deprecation warnings with MSVC
* Many portability fixes
* aperio: Detect OpenJPEG chroma subsampling breakage during `open`
* aperio: Fill in missing tiles with downsampled data
* aperio: Report MPP for slides scanned in locales with decimal comma
* hamamatsu: Support NDPI files \> 4 GB
* hamamatsu: Properly detect NDPI slides produced by NDP.toolkit
* hamamatsu: Support VMS/VMU slides without a `NoLayers` key
* hamamatsu: Report MPP for VMS/VMU
* leica: Support slides with `2010/03/10` XML namespace
* leica: Base64-decode `leica.barcode` property in `2010/10/01` namespace
* sakura: Support slides with multiple focal planes
* sakura: Support slides without tile table
* ventana: Support slides with multiple focal planes
* ventana: Improve positioning of AOIs within level
* ventana: Fix failure to recognize macro image on some slides


## Version 3.4.0, 2014-01-25

* Major internal restructuring
* New formats: Hamamatsu NDPI, Sakura SVSLIDE, Ventana BIF (preliminary)
* Add `openslide_detect_vendor()`
* Deprecate `openslide_can_open()` (not very useful and often misused)
* Document performance considerations for `openslide_open()`
* Add properties giving the bounds of the non-empty region of the slide
* leica: Support multiple main images if their levels are coplanar
* leica: Use slide size as level size
* mirax: Support PNG- and BMP-formatted slides
* mirax: Fix `Expected first 0 value` error opening some slides
* mirax: Fix incorrect tile placement on some slides without overlaps
* mirax: Never synthesize downsampled levels
* Add `OPENSLIDE_DEBUG` environment variable (`OPENSLIDE_DEBUG=?` for help)
* Fix some crashes in error paths
* Add tests for many error paths


## Version 3.3.3, 2013-04-13

* Fix inclusion of `openslide.h` with MSVC
* Properly handle Aperio JP2K slides with zero-length tiles
* Support Hamamatsu slides with blank `MacroImage` key


## Version 3.3.2, 2012-12-01

* Fix seams in MIRAX 2.2 slides (thanks, Agelos Pappas)
* Fix associated image naming in single-level Aperio slides
* Stop decoding MIRAX tiles outside requested region
* Stop decoding unneeded tiles during tile-aligned accesses
* Increase Hamamatsu VMU tile size to reduce rendering overhead
* Document performance considerations for `openslide_can_open()`


## Version 3.3.1, 2012-10-14

* Parallelize concurrent `openslide_read_region` calls on an `openslide_t`
* Eliminate background scanning of tile headers in MIRAX
* Scan many fewer tiles during first accesses to Hamamatsu VMS
* Ignore Leica Z-planes other than 0
* Add experimental tile-size properties
* Document API thread safety


## Version 3.3.0, 2012-09-08

* Support Leica SCN format (requires libtiff ≥ 4) (thanks, Agelos Pappas)
* Allow opening MIRAX 2.2 slides (though there are seams, bug #92)
* Add standardized microns-per-pixel and objective-power properties
* Add `macro` associated image in Trestle
* Rename `layer` to `level` throughout the API (deprecate `layer` functions;
  remove `layer` properties)
* Report parse errors in `openslide_open()` by returning an `openslide_t` in
  error state
* Deprecate `openslide_get_comment()`
* Add `openslide_get_version()`
* Improve command-line tools; add manpages
* Support building with MinGW-w64; drop CMake, MSVC, mingw32
* Add tests for many error paths


## Version 3.2.6, 2012-02-23

* Support downsampled MIRAX files
* Improve performance on MIRAX slides without tile overlaps
* Fix `openslide_read_region` for large dimensions on layer \> 0
  (3.2.5 regression)
* Correct subpixel error in MIRAX tile placement
* Fix unlikely use-after-free with Hamamatsu VMU


## Version 3.2.5, 2011-12-16


tools/openslide-write-associated-images.c
/*
 *  OpenSlide, a library for reading whole slide image files
 *
 *  Copyright (c) 2007-2012 Carnegie Mellon University
 *  Copyright (c) 2023 Alexandr Virodov
 *  All rights reserved.
 *
 *  OpenSlide is free software: you can redistribute it and/or modify
 *  it under the terms of the GNU Lesser General Public License as
 *  published by the Free Software Foundation, version 2.1.
 *
 *  OpenSlide is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 *  GNU Lesser General Public License for more details.
 *
 *  You should have received a copy of the GNU Lesser General Public
 *  License along with OpenSlide. If not, see
 *  <http://www.gnu.org/licenses/>.
 *
 */

#include "openslide.h"
#include "openslide-common.h"

#include <png.h>
#include <inttypes.h>
#include <glib.h>
#include <stdarg.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <setjmp.h>

static const char SOFTWARE[] = "Software";
static const char OPENSLIDE[] = "OpenSlide <https://openslide.org/>";

static void fail(const char *format, ...) G_GNUC_NORETURN;
static void fail(const char *format, ...) {
  va_list ap;

  va_start(ap, format);
  char *msg = g_strdup_vprintf(format, ap);
  va_end(ap);

  fprintf(stderr, "%s: %s\n", g_get_prgname(), msg);
  fflush(stderr);

  exit(1);
}


static void write_png(FILE *f, int64_t w, int64_t h, uint32_t* data) {
  png_structp png_ptr = png_create_write_struct(PNG_LIBPNG_VER_STRING, NULL, NULL, NULL);
  if (!png_ptr) {
    fail("Could not initialize PNG");
  }

  png_infop info_ptr = png_create_info_struct(png_ptr);
  if (!info_ptr) {
    fail("Could not initialize PNG");
  }

  if (setjmp(png_jmpbuf(png_ptr))) {
    fail("Error writing PNG");
  }

  png_init_io(png_ptr, f);

  png_set_IHDR(png_ptr, info_ptr, w, h, 8,
               PNG_COLOR_TYPE_RGB_ALPHA, PNG_INTERLACE_NONE,
               PNG_COMPRESSION_TYPE_DEFAULT,
               PNG_FILTER_TYPE_DEFAULT);

  // text
  png_text text_ptr[1];
  memset(text_ptr, 0, sizeof text_ptr);
  text_ptr[0].compression = PNG_TEXT_COMPRESSION_NONE;
  g_autofree char *key = g_strdup(SOFTWARE);
  text_ptr[0].key = key;
  g_autofree char *text = g_strdup(OPENSLIDE);
  text_ptr[0].text = text;
  text_ptr[0].text_length = strlen(text);

  png_set_text(png_ptr, info_ptr, text_ptr, 1);

  // start writing
  png_write_info(png_ptr, info_ptr);
  for (int i = 0; i < h; ++i) {
    png_write_row(png_ptr, (png_bytep) &data[w * i]);
  }

  // end
  png_write_end(png_ptr, info_ptr);
  png_destroy_write_struct(&png_ptr, &info_ptr);
}


static const struct common_usage_info usage_info = {
  "\n\tslide - to list associated images.\n\tslide associated_image output.png - to write an associated image.",
  "Write an associated image of a virtual slide to a PNG.",
};

enum mode_t {
  LIST_IMAGES,
  WRITE_IMAGE,
};

int main (int argc, char **argv) {
  enum mode_t mode;
  common_parse_commandline(&usage_info, &argc, &argv);

  switch (argc) {
    case 2: mode = LIST_IMAGES; break;
    case 4: mode = WRITE_IMAGE; break;
    default: common_usage(&usage_info); // calls exit().
  }

  const char *slide = argv[1];
  printf("Opening slide: '%s'\n", slide);

  // open slide
  g_autoptr(openslide_t) osr = openslide_open(slide);

  // check errors
  if (osr == NULL) {
    fail("%s: Not a file that OpenSlide can recognize", slide);
  }

  const char *err = openslide_get_error(osr);
  if (err) {
    fail("%s: %s", slide, err);
  }

  if (mode == LIST_IMAGES) {
    printf("Listing associated images:\n");
    const char *const *associated_image_names = openslide_get_associated_image_names(osr);
    while (*associated_image_names) {
      printf("associated image: '%s'\n", *associated_image_names);
      associated_image_names++;
    }
    printf("Done listing.\n");

  } else if (mode == WRITE_IMAGE) {
    const char *associated_image_name = argv[2];
    const char *output_file = argv[3];
    printf("Extracting associated image: '%s'\n", associated_image_name);

    int64_t w;
    int64_t h;
    openslide_get_associated_image_dimensions(osr, associated_image_name, &w, &h);
    printf("Dimensions: %"PRId64", %"PRId64"\n", w, h);

    g_autofree uint32_t *buffer = g_malloc(w * h * 4);
    openslide_read_associated_image(osr, associated_image_name, buffer);

    printf("Writing output file: '%s'\n", output_file);
    FILE* f = fopen(output_file, "wb");
    write_png(f, w, h, buffer);
    fclose(f);
  }

  // openslide_close(osr);

  return 0;
}
